### PR TITLE
Port grouping redo/v14

### PR DESCRIPTION
**Pull request by inashivb**
Previous PR: https://github.com/OISF/suricata/pull/10470

Changes since v13:
- Cleanups
- Merge the range w same SGHs early. This makes the load time a bit faster too (over and above everything w the new algo)

Still TODO:
- Cleanup of `CreatePortList` and a bug fix
- CPP check fixes
- Whitelisting fn optimization (sending RFC for that separately)

### CODE REVIEW

File: src/util-interval-tree.c [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+ * \param port Given low port
+ * \param port2 Given high port
+ * \param ptr Pointer to the node in the tree to be checked against
+ *
+ * \return true if an overlaps was found, false otherwise
+ */
+static bool IsOverlap(const uint16_t port, const uint16_t port2, const SCIntervalNode *ptr)
+{
+    /* Two intervals i and i' are said to overlap if
+     * - i (intersection) i' == NIL
```
Comment (inashivb) [inline src/util-interval-tree.c:168]: `!=` 🤦🏽‍♀️

File: src/util-interval-tree.c [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+{
+    if (a->port2 > b->max) {
+        b->max = a->port2;
+    }
+    if (a->port >= b->port) {
+        SCReturnInt(1);
+    }
+    SCReturnInt(-1);
+}
+
```
Comment (victorjulien) [inline src/util-interval-tree.c:49]: can you add
```
// cppcheck-suppress nullPointerRedundantCheck
```

File: src/detect-engine-build.c [hunk 3]
```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
         uint32_t unique_groups, int (*CompareFunc)(DetectPort *, DetectPort *));
 int CreateGroupedPortListCmpCnt(DetectPort *a, DetectPort *b);
 
+#define MAX_PORT_VAL 65536
```
Comment (victorjulien) [inline src/detect-engine-build.c:1133]: max value is 65535, right? Or UINT16_MAX (which is the same)

File: src/detect-engine-build.c [hunk 4]
```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
...
+        uint16_t port, port2;
+        UniquePortPoint *final_unique_points =
+                (UniquePortPoint *)SCCalloc(size_list, sizeof(UniquePortPoint));
+        if (final_unique_points == NULL)
+            return -1;
+        SetFinalUniquePortPoints(unique_list, size_list, final_unique_points);
+        /* Handle edge case when there is just one unique port */
+        if (size_list == 1) {
+            port = port2 = final_unique_points[0].port;
+            PISearchOverlappingPortRanges(de_ctx, port, port2, &it->tree, list);
```
Comment (victorjulien) [inline src/detect-engine-build.c:1228]: just pass `final_unique_points[0].port` directly?

File: src/detect-engine-build.c [hunk 5]
```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
...
+        } else {
+            UniquePortPoint *p1 = NULL, *p2 = NULL;
+            p1 = &final_unique_points[0];
+            p2 = &final_unique_points[1];
+            port = p1 ? p1->port : 0; // just for cppcheck
```
Comment (victorjulien) [inline src/detect-engine-build.c:1233]: declare port, port2 here?

File: src/detect-engine-build.c [hunk 6]
```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
...
+        if (final_unique_points == NULL)
+            return -1;
+        SetFinalUniquePortPoints(unique_list, size_list, final_unique_points);
+        /* Handle edge case when there is just one unique port */
+        if (size_list == 1) {
+            port = port2 = final_unique_points[0].port;
+            PISearchOverlappingPortRanges(de_ctx, port, port2, &it->tree, list);
+        } else {
+            UniquePortPoint *p1 = NULL, *p2 = NULL;
+            p1 = &final_unique_points[0];
```
Comment (victorjulien) [inline src/detect-engine-build.c:1231]: declare here

File: src/detect-engine-build.c (hunk 5 above)
Comment (victorjulien) [inline src/detect-engine-build.c:1233]: how could p1 be NULL here?

File: src/detect-engine-build.c [hunk 7]
```diff
@@ -1129,15 +1130,160 @@ int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, Detect
...
+            p2 = &final_unique_points[1];
+            port = p1 ? p1->port : 0; // just for cppcheck
+            port2 = p2->port;
+            for (uint16_t i = 1; i < size_list;) {
```
Comment (victorjulien) [inline src/detect-engine-build.c:1235]: what is the max value of `size_list`?

File: src/detect-engine-build.c (hunk 5 above)
Comment (victorjulien) [reply src/detect-engine-build.c:1233]: think both p1 and p2 can't be NULL given they point to array members

Comment (victorjulien) [reply src/detect-engine-build.c:1233]: ah I see they can be later

File: src/detect-engine-build.c (hunk 3 above)
Comment (inashivb) [reply src/detect-engine-build.c:1133]: well the name is prob wrong, I wanted to use this to allocate an array of size 65536, which I could have done by writing `UINT16_MAX + 1` 🤦🏽‍♀️

File: src/detect-engine-build.c (hunk 7 above)
Comment (inashivb) [reply src/detect-engine-build.c:1235]: 65536 if we have all the ports possible available in our ruleset

Comment (victorjulien) [reply src/detect-engine-build.c:1235]: size_list can't represent that as it has u16. Probably best to switch to int or uint32_t

**Comment by suricata-qa** [comment]
Information:
 

ERROR: QA failed on SURI_TLPW2_autofp_suri_time.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .uptime | 101 | 112 | 110.89% | 

Pipeline 18655

File: src/detect-engine-build.c (hunk 7 above)
Comment (inashivb) [reply src/detect-engine-build.c:1235]: That was stupid. Sorry. Half the places, I have taken care of the limit and half are uncaught. Thank you!

**Pull request closed without merging**